app/src/main/java/com/keylesspalace/tusky/LoginWebViewActivity.kt

```diff
@@ -0,0 +1,142 @@
+package com.keylesspalace.tusky
+
+import android.app.Activity
+import android.content.Context
+import android.content.Intent
+import android.graphics.Color
+import android.net.Uri
+import android.os.Bundle
+import android.os.Parcelable
+import android.util.Log
+import android.webkit.*
+import androidx.activity.result.contract.ActivityResultContract
+import com.keylesspalace.tusky.databinding.LoginWebviewBinding
+import com.keylesspalace.tusky.di.Injectable
+import com.keylesspalace.tusky.util.viewBinding
+import kotlinx.parcelize.Parcelize
+
+/** Contract for starting [LoginWebViewActivity]. */
+class OauthLogin : ActivityResultContract<LoginData, LoginResult>() {
+    override fun createIntent(context: Context, input: LoginData): Intent {
+        val intent = Intent(context, LoginWebViewActivity::class.java)
+        intent.putExtra(DATA_EXTRA, input)
+        return intent
+    }
+
+    override fun parseResult(resultCode: Int, intent: Intent?): LoginResult {
+        // Can happen automatically on up or back press
+        return if (resultCode == Activity.RESULT_CANCELED) {
+            LoginResult.Cancel
+        } else {
+            intent!!.getParcelableExtra(RESULT_EXTRA)!!
+        }
+    }
+
+    companion object {
+        private const val RESULT_EXTRA = "result"
+        private const val DATA_EXTRA = "data"
+
+        fun parseData(intent: Intent): LoginData {
+            return intent.getParcelableExtra(DATA_EXTRA)!!
+        }
+
+        fun makeResultIntent(result: LoginResult): Intent {
+            val intent = Intent()
+            intent.putExtra(RESULT_EXTRA, result)
+            return intent
+        }
+    }
+}
+
+@Parcelize
+data class LoginData(
+    val url: Uri,
+    val oauthRedirectUrl: Uri,
+) : Parcelable
+
+sealed class LoginResult : Parcelable {
+    @Parcelize
+    data class Ok(val code: String) : LoginResult()
+
+    @Parcelize
+    data class Err(val errorMessage: String) : LoginResult()
+
+    @Parcelize
+    object Cancel : LoginResult()
+}
+
+/** Activity to do Oauth process using WebView. */
+class LoginWebViewActivity : BaseActivity(), Injectable {
+    private val binding by viewBinding(LoginWebviewBinding::inflate)
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        val data = OauthLogin.parseData(intent)
+
+        setContentView(binding.root)
+
+        setSupportActionBar(binding.loginToolbar)
+        supportActionBar?.setDisplayHomeAsUpEnabled(true)
+        supportActionBar?.setDisplayShowTitleEnabled(false)
+
+        val webView = binding.loginWebView
+        webView.settings.allowContentAccess = false
+        webView.settings.allowFileAccess = false
+        webView.settings.databaseEnabled = false
+        webView.settings.displayZoomControls = false
+        webView.settings.javaScriptCanOpenWindowsAutomatically = false
+        webView.settings.userAgentString += "Tusky ${BuildConfig.VERSION_NAME}"
```

Comment:
maybe add the Android Version like here https://github.com/tuskyapp/Tusky/blob/develop/app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt#L81

Reply:
Good call!

Reply:
I now injected exactly the same UA but then
 - Default UA has a lot of useful info e.g. engine version
 - Default UA already has Android version in it
 - We don't need OkHttp version for this UA

But I've changed a format now to be closer to what UA should be